### PR TITLE
Fix: Change is function signature check

### PR DESCRIPTION
**Pull request by BackEndTea**
This PR:

* [x] Removes `isFunctionBodyMutator` & `isFunctionSignatureMutator`
* [x] Removes the `FunctionBodyMutator` class & the `FunctionSignatureMutator` class
* [x] Uses the AST to determine if a node is on the function signature

This change causes the MSI to go up, but the Covered Code MSI to go down.

Related to: #101

### CODE REVIEW

File: src/TestFramework/Coverage/CodeCoverageData.php [hunk 1]
```diff
@@ -199,6 +193,18 @@ private function addTestExecutionInfo(array $coverage): array
         return $newCoverage;
     }
 
+    public function isLineFunctionSignature(string $filePath, int $line): bool
```
Comment (BackEndTea) [inline src/TestFramework/Coverage/CodeCoverageData.php]: Since this gets called multiple times it might be beneficial to 'cache' its results somehow.

File: src/Visitor/MutationsCollectorVisitor.php [hunk 2]
```diff
@@ -58,19 +58,20 @@ public function __construct(array $mutators, string $filePath, array $fileAst, C
     public function leaveNode(Node $node)
     {
         foreach ($this->mutators as $mutator) {
-            if ($mutator->isFunctionBodyMutator() &&
-                !$node->getAttribute(WrappedFunctionInfoCollectorVisitor::IS_INSIDE_FUNCTION_KEY)) {
-                continue;
+            if (!$this->codeCoverageData->isLineFunctionSignature($this->filePath, $node->getLine())) {
```
Comment (sidz) [inline src/Visitor/MutationsCollectorVisitor.php]: I'd say that we can save result into some variable as we're executing it at least 3 times

File: src/TestFramework/Coverage/CodeCoverageData.php [hunk 3]
```diff
@@ -199,6 +193,18 @@ private function addTestExecutionInfo(array $coverage): array
...
+    {
+        $coverage = $this->getCoverage();
+        foreach ($coverage[$filePath]['byMethod'] as $method => $coverageInfo) {
+            if ($line == $coverageInfo['startLine']) {
```
Comment (sidz) [inline src/TestFramework/Coverage/CodeCoverageData.php]: could we use strict types here `$line === (int) $coverageInfo['startLine']` ?

File: tests/Mutant/Generator/MutationsGeneratorTest.php [hunk 4]
```diff
@@ -58,26 +50,12 @@ public function test_it_can_skip_not_covered_on_file_level()
         $this->assertCount(0, $mutations);
     }
 
-    public function test_it_can_skip_not_covered_on_file_line_level()
```
Comment (borNfreee) [inline tests/Mutant/Generator/MutationsGeneratorTest.php]: could you please explain why did you remove it?

**Comment by BackEndTea** [comment]
I have applied received feedback @sidz & @borNfreee 

Before this gets merged i'd like to add a few tests for `isLineFunctionSignature`.

An optimisation idea would be to mark all mutations that could *possibly* be on the function signature and only check if it is on a signature for those.

And instead of looping over the methods for every mutation it might be better to save that info as a hash-map or something like that.

File: tests/TestFramework/Coverage/CodeCoverageDataTest.php [hunk 5]
```diff
@@ -174,6 +178,58 @@ public function test_it_throws_an_exception_when_no_coverage_found()
         $coverage->hasTests('/abc/def.php');
     }
 
+    /**
+     * @dataProvider providesFunctionSignatureLineNumbers
+     */
+    public function test_it_correctly_finds_that_line_is_signature($lineNumber)
```
Comment (borNfreee) [inline tests/TestFramework/Coverage/CodeCoverageDataTest.php]: `int $lineNumber`

File: tests/TestFramework/Coverage/CodeCoverageDataTest.php [hunk 6]
```diff
@@ -174,6 +178,58 @@ public function test_it_throws_an_exception_when_no_coverage_found()
...
+     * @dataProvider providesNonFunctionSignatureLineNumbers
+     */
+    public function test_it_knows_when_lines_are_not_function_signature($lineNumber)
+    {
+        $coverageData = $this->getCodeCoverageData();
+        $filePath = '/tests/Fixtures/Files/phpunit/coverage-xml/FirstLevel/firstLevel.php';
+
+        //Known function signature lines
+        $this->assertFalse(
+            $coverageData->isLineFunctionSignature($filePath,$lineNumber),
```
Comment (borNfreee) [inline tests/TestFramework/Coverage/CodeCoverageDataTest.php]: space after comma.
I've added `tests` folder to PHPCSFixer config and fixed all issues, could you please rebase?

File: tests/TestFramework/Coverage/CodeCoverageDataTest.php [hunk 7]
```diff
@@ -174,6 +178,58 @@ public function test_it_throws_an_exception_when_no_coverage_found()
...
+
+    /**
+     * @dataProvider providesNonFunctionSignatureLineNumbers
+     */
+    public function test_it_knows_when_lines_are_not_function_signature($lineNumber)
+    {
+        $coverageData = $this->getCodeCoverageData();
+        $filePath = '/tests/Fixtures/Files/phpunit/coverage-xml/FirstLevel/firstLevel.php';
+
+        //Known function signature lines
```
Comment (borNfreee) [inline tests/TestFramework/Coverage/CodeCoverageDataTest.php]: incorrect comment (should be opposite)

**Comment by BackEndTea** [comment]
Rebased and addressed  issues

**Comment by borNfreee** [comment]
Everything looks good.

FYI: I tried to run this branch on infection itself, and there are 918 mutations while master branch has 1010+, so 100 mutations difference. That's suspicious. 

I ran it on a small project and found that `PublicVisibility` mutator just ignored in some cases, didn't manage to find a bug yet as I'm a bit out of time today. Will continue debugging tomorrow.

**Comment by BackEndTea** [comment]
>FYI: I tried to run this branch on infection itself, and there are 918 mutations while master branch has 1010+, so 100 mutations difference. That's suspicious.

>I ran it on a small project and found that PublicVisibility mutator just ignored in some cases, didn't manage to find a bug yet as I'm a bit out of time today. Will continue debugging tomorrow.

Thats strange, i'll take a look at that as well.

>Also, seems like tests are broken atm, `test_it_can_skip_not_covered_on_file_line_level` method should contain 14 instead of 13 because of added new empty line after CS fixes in master.

You're right, i fixed it now.

**Comment by BackEndTea** [comment]
I think the issue lies somewhere in this function: 
https://github.com/infection/infection/blob/11c0f87f0f3061c955da91ba5b3c266281ba18bd/src/TestFramework/PhpUnit/Coverage/CoverageXmlParser.php#L57-L92

Right now if 'byMethod' is empty the `isLineFunctionSignature` thinks the function is not a signature, which may cause an issue in the `leaveNode` function

**Comment by borNfreee** [comment]
Finally, found a bug. We can't rely on the code coverage in `isLineFunctionSignature()` function because if the method from the source code has coverage 0% (was not executed at all), PHPUnit's Code Coverage does not have information in xml reports about it and we can't determine if particular line is a line of the function signature.

Moreover, this is not a correct approach since such one-line function breaks the idea:

```php
public function foo() { doSomething(); return true }
```

In the example above, each node _"is on the functon signature line"_, but what we actually whant to know is whether the node is **part** of the function signature, not on its line.

So, I tried to get that information from the AST. My manual tests show everything works the same as in master.

Could you please test this branch with your code to see if it fixes the original issue @BackEndTea ?

File: src/Visitor/WrappedFunctionInfoCollectorVisitor.php [hunk 8]
```diff
@@ -48,6 +50,24 @@ public function leaveNode(Node $node)
...
+            return true;
+        }
+
+        if (!$node->hasAttribute(ParentConnectorVisitor::PARENT_KEY)) {
+            return false;
+        }
+
+        $parent = $node->getAttribute(ParentConnectorVisitor::PARENT_KEY);
+
+        if ($parent instanceof Node\Param) {
```
Comment (sidz) [inline src/Visitor/WrappedFunctionInfoCollectorVisitor.php]: `return $parent instanceof Node\Param;`

**Comment by BackEndTea** [comment]
Your change was in the right direction, but it did not fix my issues.

This change should do the trick

File: src/Visitor/WrappedFunctionInfoCollectorVisitor.php [hunk 9]
```diff
@@ -36,6 +36,8 @@ public function enterNode(Node $node)
         if ($this->isFunctionLikeNode($node)) {
             $this->scopeStack[] = $node;
             $node->setAttribute(self::IS_ON_FUNCTION_SIGNATURE, true);
+        }elseif ($this->isPartOfFunctionSignature($node)) {
```
Comment (borNfreee) [inline src/Visitor/WrappedFunctionInfoCollectorVisitor.php]: Seems like all items that "are part of function signature", e.g. return value, will never get `FUNCTION_SCOPE_KEY` attribute (the next `elseif` section. This is a BC

**Comment by BackEndTea** [comment]
I rebased against master since that had some changes in the `WrappedFunctionInfoCollectorVisitor`.

> Seems like all items that "are part of function signature", e.g. return value, will never get FUNCTION_SCOPE_KEY attribute (the next elseif section. This is a BC break

would 
```php
    public function enterNode(Node $node)
    {
        $isInsideFunction = $this->isInsideFunction($node);

        if ($isInsideFunction) {
            $node->setAttribute(self::IS_INSIDE_FUNCTION_KEY, true);
        }

        if ($this->isPartOfFunctionSignature($node)) {
            $node->setAttribute(self::IS_ON_FUNCTION_SIGNATURE, true);
        }

        if ($this->isFunctionLikeNode($node)) {
            $this->scopeStack[] = $node;
            $node->setAttribute(self::IS_ON_FUNCTION_SIGNATURE, true);
        } elseif ($isInsideFunction) {
            $node->setAttribute(self::FUNCTION_SCOPE_KEY, $this->scopeStack[count($this->scopeStack) - 1]);
        }
    }
```
Do the trick?
I'm not 100% on what the different attributes or the `scopeStack` do

**Comment by borNfreee** [comment]
> Would [...] Do the trick?

yes. I've added more tests and fixed a couple of minor issues.

**Comment by borNfreee** [comment]
Tested original issue on you Array-Meta repository. 100% for all MSI metrics 👍 

Thank you @BackEndTea

**Comment by BackEndTea** [comment]
Thank you @borNfreee & @sidz for merging and for all the feedback.
